### ISSUE TITLE
New Loss and new metric for segmentation task

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#77]
<html>
<body>
<html><head></head><body><p>For your current color/enhancement tasks (brightness, contrast, gamma, saturation), MSE is the right optimization loss — keep it.</p>
<p>For binary mask segmentation, you need changes in <strong>both</strong> places:</p>
<p><strong>1. Optimization loss</strong> — MSE works but isn't ideal for masks. Better options:</p>

Loss | Why
-- | --
BCE (Binary Cross-Entropy) | Standard for binary classification per-pixel. Penalizes confident wrong predictions heavily.
Dice Loss | Directly optimizes overlap (IoU-like). Handles class imbalance well (small object, large background).
BCE + Dice | Best of both — BCE for pixel accuracy, Dice for region overlap. Most common combo.

<p>MSE treats a 0.9 prediction vs 1.0 target the same as 0.1 vs 0.2 — it doesn't understand "this is a classification problem." BCE and Dice do.</p>
<p><strong>2. Evaluation metrics</strong> — PSNR/SSIM are meaningless for binary masks. Use:</p>
<ul>
<li><strong>IoU (Jaccard)</strong> — intersection over union, the standard</li>
<li><strong>Dice score</strong> — similar to IoU, commonly reported</li>
<li><strong>Pixel accuracy</strong> — simple but misleading with class imbalance</li>
</ul>
<p><strong>Recommendation:</strong> Change both. Use <strong>BCE + Dice</strong> as optimization loss, evaluate with <strong>IoU + Dice</strong>. Keep MSE/PSNR/SSIM for your enhancement tasks — just branch on task type.</p></body></html>
</body>
</html>

**Comment by xp1632** [comment]
Two clean options:

**Option A: Pass loss function to the optimizer (simplest)**

```python
# Caller decides the loss
optimizer.optimize(
    workflow=workflow,
    input_images=inputs,
    target_images=targets,
    loss_fn=torch.nn.BCELoss()  # or torch.nn.MSELoss() for enhancement
)
```

The optimizer doesn't need to know the task — it just uses whatever loss it's given. `ProgramEvaluator` already stores `self.loss_fn`, so you'd just pass it through instead of hardcoding `MSELoss()`.

**Option B: Pass a task type string**

```python
optimizer.optimize(..., task="segmentation")
```

Then internally map `"segmentation" → BCELoss + DiceLoss`, `"enhancement" → MSELoss`. Simpler API but less flexible.

**I'd go with Option A** — it keeps the optimizer generic and the caller (your evaluation script) is already the one that knows the task type. The change is minimal: add a `loss_fn` parameter that defaults to `MSELoss()` for backward compatibility, and thread it through to `ProgramEvaluator`.
